cmake/ext_project_autotools.cmake

```diff
@@ -0,0 +1,121 @@
+# CONFIGURE - options for configure script.
+# BYPRODUCTS - products caller targets depend upon.
+#
+# Note that incremental rebuild in case autotools files or config files are
+# changed is available only if CMake version is at least 3.12. Otherwise
+# rebuild will not be triggered. Rebuild is only required for developers
+# which presumably has newer CMake versions.
+function(ext_project_autotools name)
+    set(sv_keywords
+        DIR
+    )
+    set(mv_keywords
+        CONFIGURE
+        BYPRODUCTS
+    )
+    set(options
+        ""
+    )
+    cmake_parse_arguments(
+        ARGS "${options}" "${sv_keywords}" "${mv_keywords}" ${ARGN}
+    )
+    if (NOT DEFINED ARGS_DIR)
+        message(FATAL_ERROR "DIR argument is mandatory")
+    endif()
+
+    # CONFIGURE_DEPENDS available since 3.12
+    # VERSION_GREATER_EQUAL is only since 3.7
+    if (${CMAKE_VERSION} VERSION_GREATER "3.12" OR
+        ${CMAKE_VERSION} VERSION_EQUAL "3.12")
+        set(autotool_files_standard
+            configure.ac
+            acinclude.m4
+        )
+        list_add_prefix(
+            autotool_files_standard "${ARGS_DIR}/" autotool_files_standard
+        )
+
+        file(
+            GLOB_RECURSE autotool_files_am
+            CONFIGURE_DEPENDS
+            RELATIVE "${CMAKE_SOURCE_DIR}"
+            "${ARGS_DIR}/*.am"
+        )
+
+        set(autotool_files ${autotool_files_am} ${autotool_files_standard})
+
+        file(
+            GLOB_RECURSE config_files_in
+            CONFIGURE_DEPENDS
+            RELATIVE "${CMAKE_SOURCE_DIR}"
+            "${ARGS_DIR}/*.in"
+        )
+    endif()
+
+    list_add_prefix(ARGS_BYPRODUCTS "${ARGS_DIR}/" byproducts)
+
+    file(MAKE_DIRECTORY ${CMAKE_BINARY_DIR}/${ARGS_DIR})
+
+    add_custom_command(
+        OUTPUT ${CMAKE_SOURCE_DIR}/${ARGS_DIR}/configure
+        WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}/${ARGS_DIR}
+        COMMAND
+            autoreconf -i
+        COMMAND
+            # autoreconf do not touch configure on all paths. For example if
+            # only *.am are changed. We don't aim to follow precise dependency
+            # path of autotools thus let's just update configure in this case.
+            touch configure
+        DEPENDS ${autotool_files}
+    )
+
+    add_custom_command(
+        OUTPUT ${ARGS_DIR}/Makefile
+        WORKING_DIRECTORY ${ARGS_DIR}
+        COMMAND ${CMAKE_SOURCE_DIR}/${ARGS_DIR}/configure ${ARGS_CONFIGURE}
+        DEPENDS ${ARGS_DIR}/configure
+    )
+
+    # Use artificial timestamp file to make a dependency with main target.
+    # We can't use just ${config_files_in} counterparts because
+    # depending on configure options not all counterparts could be generated
+    # by config.status and as a result it will be rerun unconditionally.
+    add_custom_command(
+        OUTPUT
+            ${ARGS_DIR}/config.stamp.tarantool
+        WORKING_DIRECTORY ${ARGS_DIR}
+        COMMAND ./config.status
+        COMMAND touch config.stamp.tarantool
```

Comment:
I usually use in-source build (`cmake . <...> && make -j`) and the `config.stamp.tarantool` file is generated in `third_party/libunwind`. In this case it causes the submodule to be marked as containing untracked content:

> modified:   third_party/libunwind (untracked content)

We can either choose a gitignored directory within libunwind's sources (say, `config`) or one in tarantool sources (say, `build`).

Reply:
Fortunately `config.log` timestamp is updated on `config.status` run thus I don't need artificial `config.stamp.tarantool`. `config.log` is in libunwind's `.gitignore` and other other sane third patry projects will have it in `.gitignore` too probably.

So modified this place the next way:
```diff
--- a/cmake/ext_project_autotools.cmake
+++ b/cmake/ext_project_autotools.cmake
@@ -84,20 +84,23 @@ function(ext_project_autotools name)
         DEPENDS ${ARGS_DIR}/configure
     )

     # Use artificial timestamp file to make a dependency with main target.
     # We can't use just ${config_files_in} counterparts because
     # depending on configure options not all counterparts could be generated
     # by config.status and as a result it will be rerun unconditionally.
     add_custom_command(
         OUTPUT
-            ${ARGS_DIR}/config.stamp.tarantool
+            ${ARGS_DIR}/config.log
         WORKING_DIRECTORY ${ARGS_DIR}
         COMMAND ./config.status
-        COMMAND touch config.stamp.tarantool
         DEPENDS
             ${config_files_in}
             # Actually we depend on config.status but it updates
             # it's own timestamp on every run.
             ${ARGS_DIR}/Makefile
     )

@@ -114,7 +117,7 @@ function(ext_project_autotools name)
             ${make}
         DEPENDS
             ${ARGS_DIR}/Makefile
-            ${ARGS_DIR}/config.stamp.tarantool
+            ${ARGS_DIR}/config.log
         BYPRODUCTS
             ${byproducts}
     )
```